### PR TITLE
BUG: Fix interval constructor to handle left=right case

### LINKED ISSUE #26893: interval.overlaps() mishandles empty intervals
The special case where `left=right` is not checked properly by the constructor, only `closed='both'` (a point)  or `closed='neither'` (an empty interval) makes sense but '[0,0)' or `(0,0]` are nonsensical. The `overlaps` method returns inconsistent results when encountering such Intervals. 

```python
import pandas as pd
pd.Interval(0,0,'left')
```

**Update:**
Examples (from below):
```python
In [4]:  
   ...: import pandas as pd
   ...: from pandas import Interval
   ...: 
   ...: a=Interval(0,1,'both')
   ...: b=Interval(0,0,'right')
   ...: print(a.overlaps(b))
   ...: b=Interval(0,0,'left')
   ...: print(a.overlaps(b))
True
False
```

### PR DESCRIPTION
**Pull request by ghost**
- [X] closes #26893
- [] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

**Update:**

[ ] - Following https://github.com/pandas-dev/pandas/pull/26894#pullrequestreview-250742121, needs tests about creating an `Intervalindex` with '(x,x]', also checking that the repr doesn't blowup.

### CODE REVIEW

**Comment by pilkibun** [comment]
Ready.

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -617,13 +617,12 @@ Strings
...
 -
 
-
 Interval
 ^^^^^^^^
 
 - Construction of :class:`Interval` is restricted to numeric, :class:`Timestamp` and :class:`Timedelta` endpoints (:issue:`23013`)
 - Fixed bug in :class:`Series`/:class:`DataFrame` not displaying ``NaN`` in :class:`IntervalIndex` with missing values (:issue:`25984`)
--
+- :class:Interval will complain when creating point interval with an inconsistent `closed` value (:issue:`26893`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this is not a very friendly comment, can you re-word, put double backticks around ``closed=`. This is also not clear what you are actually fixing here. What is an inconsistent 'closed' value?

File: pandas/tests/arrays/interval/test_interval.py [hunk 2]
```diff
@@ -65,3 +65,23 @@ def test_repr_matches():
     a = repr(idx)
     b = repr(idx.values)
     assert a.replace("Index", "Array") == b
+
+
+def test_point_interval():
```
Comment (jreback) [inline pandas/tests/arrays/interval/test_interval.py]: make the error conditions as a separate parmeterized test

File: pandas/tests/arrays/interval/test_interval.py [hunk 3]
```diff
@@ -65,3 +65,23 @@ def test_repr_matches():
...
+    match = "both/neither sides must be closed when left == right"
+    for closed in ('left', 'right'):
+        with pytest.raises(ValueError, match=match):
+            pd.Interval(0, 0, closed)
+
+    pd.Interval(0, 0, 'neither')  # no exception
```
Comment (jreback) [inline pandas/tests/arrays/interval/test_interval.py]: assert the result value

File: pandas/tests/arrays/interval/test_interval.py [hunk 4]
```diff
@@ -65,3 +65,23 @@ def test_repr_matches():
...
+    pd.Interval(0, 0, 'both')  # no exception
+
+    assert not pd.Interval(0, 1, "left").overlaps(pd.Interval(0, 0, "neither"))
```
Comment (jreback) [inline pandas/tests/arrays/interval/test_interval.py]: would prefer this is a paramterized test as this is pretty hard to read this way

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/tile.py [hunk 5]
```diff
@@ -476,9 +476,14 @@ def _format_labels(bins, precision, right=True,
     if right and include_lowest:
         # we will adjust the left hand side by precision to
         # account that we are all right closed
-        v = adjust(labels[0].left)
-
-        i = IntervalIndex([Interval(v, labels[0].right, closed='right')])
+        #
+        # We cannot access labels[0].left because that
+        # calls Interval(left, right, closed='right')
```
Comment (jreback) [inline pandas/core/reshape/tile.py]: I am not clear what the issue is here, can you show a full example

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (pilkibun) [reply doc/source/whatsnew/v0.25.0.rst]: I'm talking about `(0,0]` or `[0,0)` when the same point is both open and closed, which makes no sense. Can you suggest a better wording?

File: pandas/tests/arrays/interval/test_interval.py (hunk 3 above)
Comment (pilkibun) [reply pandas/tests/arrays/interval/test_interval.py]: `assert x==x`???

File: pandas/tests/arrays/interval/test_interval.py (hunk 2 above)
Comment (pilkibun) [reply pandas/tests/arrays/interval/test_interval.py]: done

File: pandas/tests/arrays/interval/test_interval.py (hunk 4 above)
Comment (pilkibun) [reply pandas/tests/arrays/interval/test_interval.py]: done

File: pandas/core/reshape/tile.py (hunk 5 above)
Comment (pilkibun) [reply pandas/core/reshape/tile.py]: removed.

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (pilkibun) [reply doc/source/whatsnew/v0.25.0.rst]: Reworded with explicit cases

File: doc/source/whatsnew/v0.25.0.rst [hunk 6]
```diff
@@ -617,13 +617,12 @@ Strings
...
 -
 
-
 Interval
 ^^^^^^^^
 
 - Construction of :class:`Interval` is restricted to numeric, :class:`Timestamp` and :class:`Timedelta` endpoints (:issue:`23013`)
 - Fixed bug in :class:`Series`/:class:`DataFrame` not displaying ``NaN`` in :class:`IntervalIndex` with missing values (:issue:`25984`)
--
+- :class:Interval now refuses to create intervals like `[0,0)` or `(0,0]` (:issue:`26893`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.0.rst]: - `` :class:`Interval` ``
-  This description is a bit too colloquial. Could you describe the conditions where the new `ValueError` is raised.

**Comment by TomAugspurger** [comment]
See https://github.com/pandas-dev/pandas/issues/26893#issuecomment-502846281. We may not be progressing with this.

**Review by jschendel — CHANGES_REQUESTED** [review]
I'm -1 on this as detailed in https://github.com/pandas-dev/pandas/issues/26893#issuecomment-502846281.  If my -1 is overruled, then these changes would also need to be propagated to the construction of `IntervalArray` and `IntervalIndex`.  Currently you can create and operate on arrays/indexes containing an empty interval, only to get an error message when the code actually tries to materialize a scalar interval:
```python
In [1]: import pandas as pd; pd.__version__
Out[1]: '0.25.0.dev0+757.gb3d18ba887'

In [2]: ia = pd.arrays.IntervalArray.from_arrays([1,2], [1, 2])

In [3]: ia.mid
Out[3]: Float64Index([1.0, 2.0], dtype='float64')

In [4]: ia.length
Out[4]: Int64Index([0, 0], dtype='int64')

In [5]: ia[0]
---------------------------------------------------------------------------
ValueError: both/neither sides must be closed when left == right
```

**Comment by pilkibun** [comment]
@jschendel, I've seen your comment, lets keep the discussion all in one place in #26893. 

Absolutely, this should be fixed. I haven't looked at it yet, but it just might be possible that the exception is raised during specifically repr creation for formatting purposes, in which case this might be solved by silently converting to closed='neither' on creation.

**Comment by pilkibun** [comment]
closed in favor of #26937

**Comment by pilkibun** [comment]
While writing the new PR, I came across `IntervalTree` for the first time. It probably relies on intervals being orderable, right? something to keep in mind if ever reconsidering the issue of a singleton.

**Pull request closed without merging**